### PR TITLE
Date and year

### LINKED ISSUE #2: prepare structure for next years
This year's data should remain isolated but still accessible in the next years.
Though the individual years could still have statical pages, I believe it'll be simpler to maintain and allows individuall customization each year,

### PR DESCRIPTION
**Pull request by Kenedy**
It would be good if you can test those validations. I had to redesign them a bit and because we are not using any robust framework there and photo upload and date has to be treated differently, they are bit clumsy.

### CODE REVIEW

File: routes/uploadResult.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ const fs = require('fs');
 const path = require('path');
 const nanoid = require('nanoid').nanoid;
 const photosDir = 'public/images/photos';
+const minDateTick = Date.parse('2010-01-10');
```
Comment (jpl-hp69) [inline routes/uploadResult.js]: 2021

File: public/registration.html [hunk 2]
```diff
@@ -35,6 +36,23 @@ <h2>Registrace výsledků</h2>
             </div>
         </div>
 
+        <div class="row" data-bind="visible: showDateRequired">
+            <div class="col-25"></div>
+            <div class="col-75 validation-error">Datum je povinné</div>
+        </div>
+        <div class="row" data-bind="visible: showDateInvalid">
+            <div class="col-25"></div>
+            <div class="col-75 validation-error">Datum není v pořádku. Nemůže být před 10.1.2020 a nemůže být ani v budoucnu.</div>
```
Comment (jpl-hp69) [inline public/registration.html]: asi spíš 2021

File: public/registration.html [hunk 3]
```diff
@@ -35,6 +36,23 @@ <h2>Registrace výsledků</h2>
...
+        </div>
+        <div class="row">
+            <div class="col-25">
+                <label for="date">Datum:</label>
+            </div>
+            <div class="col-75">
+                <input type="date" id="date" name="date" data-bind="value: date, event: { input: onDateChange }">
```
Comment (jpl-hp69) [inline public/registration.html]: the date should be pre-filled with "today"

**Comment by jpl-hp69** [comment]
my current (local) results disappeared from the list of results - just the name is still there. I also do not see the uploaded results any more.

File: public/javascripts/registration.js [hunk 4]
```diff
@@ -1,18 +1,62 @@
...
+        return model.name().length > 0;
+    };
     model.showNameRequired = ko.pureComputed(function () {
-        return model.nameTouched() && model.name().length === 0;
-    }),
-    model.onPhotoChange = function (data, event) {
+        return model.nameTouched() && !model.isNameValid();
+    });
+    model.onDateChange = function () {
+        model.dateTouched(true);
```
Comment (jpl-hp69) [inline public/javascripts/registration.js]: the date is not pre-filled so there is no change on submit and the (server-side) error is displayed first after the upload

**Comment by Kenedy** [comment]
> my current (local) results disappeared from the list of results - just the name is still there. I also do not see the uploaded results any more.

You have to clear previous data from results.json (and you can delete those pictures too). As previous data version is missing date, the script is failing and page looks empty.

File: public/javascripts/registration.js [hunk 5]
```diff
@@ -1,18 +1,62 @@
...
+    };
+    model.dateFormatted = ko.pureComputed(function () {
+        return model.isDateValid()
+            ? moment(model.date()).format('D. M. YYYY')
+            : null;
+    });
+    model.showDateRequired = ko.pureComputed(function () {
+        return model.dateTouched() && !model.date();
+    });
+    model.isDateValid = function () {
```
Comment (jpl-hp69) [inline public/javascripts/registration.js:30]: do we have an onDateChanged --> update preview?

File: public/registration.html [hunk 6]
```diff
@@ -49,7 +49,7 @@ <h2>Registrace výsledků</h2>
                 <label for="date">Datum:</label>
             </div>
             <div class="col-75">
-                <input type="date" id="date" name="date" data-bind="value: date, event: { input: onDateChange }">
+                <input type="date" id="date" name="date" data-bind="value: date, event: { input: onDateChange }, attr: { min: '2021-01-10', max: moment().format('YYYY-MM-DD') }">
```
Comment (jpl-hp69) [inline public/registration.html]: so is it moment() or moment(new Date())

Comment (Kenedy) [reply public/registration.html]: It is both the same, parameter is optional, current date and time are used if it is missing. https://momentjs.com/

File: public/javascripts/registration.js [hunk 7]
```diff
@@ -5,7 +5,7 @@ $(document).ready(function () {
     var model = {
         name: ko.observable(),
         nameTouched: ko.observable(false),
-        date: ko.observable(),
+        date: ko.observable(moment(new Date()).format('YYYY-MM-DD')),
```
Comment (jpl-hp69) [inline public/javascripts/registration.js:8]: date: ko.observable(moment().format('YYYY-MM-DD')),

File: public/registration.html (hunk 2 above)
Comment (jpl-hp69) [reply public/registration.html]: still needs fix

**Pull request merged**